build/build.go

```diff
@@ -630,50 +639,58 @@ func (s *Session) BuildPackage(pkg *PackageData) (*compiler.Archive, error) {
 			if importedPkgPath == "unsafe" || ignored {
 				continue
 			}
-			importedPkg, _, err := s.buildImportPathWithSrcDir(importedPkgPath, pkg.Dir)
+			_, importedArchive, err := s.buildImportPathWithSrcDir(importedPkgPath, pkg.Dir)
 			if err != nil {
 				return nil, err
 			}
-			impModTime := importedPkg.SrcModTime
-			if impModTime.After(pkg.SrcModTime) {
-				pkg.SrcModTime = impModTime
-			}
+
+			fmt.Fprintf(pkgHash, "import: %v\n", importedPkgPath)
+			fmt.Fprintf(pkgHash, "  hash: %#x\n", importedArchive.Hash)
 		}
 
 		for _, name := range append(pkg.GoFiles, pkg.JSFiles...) {
-			fileInfo, err := statFile(filepath.Join(pkg.Dir, name))
+			fp := filepath.Join(pkg.Dir, name)
+			file, err := s.bctx.OpenFile(fp)
 			if err != nil {
-				return nil, err
-			}
-			if fileInfo.ModTime().After(pkg.SrcModTime) {
-				pkg.SrcModTime = fileInfo.ModTime()
+				return nil, fmt.Errorf("failed to open %v: %v", fp, err)
 			}
+			fmt.Fprintf(pkgHash, "file: %v\n", fp)
+			n, _ := io.Copy(pkgHash, file)
+			file.Close()
+			fmt.Fprintf(pkgHash, "%d bytes\n", n)
 		}
 
-		pkgObjFileInfo, err := os.Stat(pkg.PkgObj)
-		if err == nil && !pkg.SrcModTime.After(pkgObjFileInfo.ModTime()) {
-			// package object is up to date, load from disk if library
-			pkg.UpToDate = true
-			if pkg.IsCommand() {
-				return nil, nil
-			}
+		// no commands are archived
+		if pkg.IsCommand() {
+			goto CacheMiss
+		}
 
-			objFile, err := os.Open(pkg.PkgObj)
-			if err != nil {
-				return nil, err
+		objFile, err := os.Open(pkg.PkgObj)
+		if err != nil {
+			if os.IsNotExist(err) {
+				goto CacheMiss
 			}
-			defer objFile.Close()
+			return nil, err
+		}
 
-			archive, err := compiler.ReadArchive(pkg.PkgObj, pkg.ImportPath, objFile, s.Types)
-			if err != nil {
-				return nil, err
-			}
+		archive, err := compiler.ReadArchive(pkg.PkgObj, pkg.ImportPath, objFile, s.Types)
+		objFile.Close()
```

Comment:
Why is `objFile` closed here?

Reply:
Because we end up opening a number of files as part of this hash calculation (recursive calls), I wanted to be clearer on when those open files will be closed.

Reply:
I've just pushed up a commit that refactors to use `defer`; that way it's more consistent with other places, I agree.